### PR TITLE
Add HPOS compat queries for tracker.

### LINKED ISSUE #38148: [COT] Make WC_Tracker compatible
`WC_Tracker` queries some order data and does not seem to be aware of HPOS. [Example:](https://github.com/woocommerce/woocommerce/blob/e7d67710fdbee8d0126adc6235d5fa4574ac7310/plugins/woocommerce/includes/class-wc-tracker.php#L509-L530)

```php
$orders_by_gateway = $wpdb->get_results(
	"
	SELECT
		gateway, currency, SUM(total) AS totals, COUNT(order_id) AS counts
	FROM (
		SELECT
			orders.id AS order_id,
			MAX(CASE WHEN meta_key = '_payment_method' THEN meta_value END) gateway,
			MAX(CASE WHEN meta_key = '_order_total' THEN meta_value END) total,
			MAX(CASE WHEN meta_key = '_order_currency' THEN meta_value END) currency
		FROM
			{$wpdb->prefix}posts orders
		LEFT JOIN
			{$wpdb->prefix}postmeta order_meta ON order_meta.post_id = orders.id
		WHERE orders.post_type = 'shop_order'
			AND orders.post_status in ( 'wc-completed', 'wc-processing', 'wc-refunded' )
			AND meta_key in( '_payment_method','_order_total','_order_currency')
		GROUP BY orders.id
	) order_gateways
	GROUP BY gateway, currency
	"
);
```

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds compatible queries for HPOS in WC_Tracker class so that we continue to get accurate data as expected.

Closes #38148 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. With posts as authoritative data store and sync disabled, generate about 100 orders using [WC Smooth Generator](https://github.com/woocommerce/wc-smooth-generator).
2. On a shell, run the following command: `WC_Tracker::get_tracking_data()['orders'];`. Copy paste the result somewhere.
3. Run the migration using the CLI command: `wp wc cot sync'`
4. Once migration finishes, run the shell command again: `WC_Tracker::get_tracking_data()['orders'];`. Compare this result with the result from step 2. It should be same.

### CODE REVIEW

**Comment by vedanshujain** [comment]
Note that CI may fail because of intermittent npm issues: https://status.npmjs.org/ . Please rebuild once status site is green.

File: plugins/woocommerce/includes/class-wc-tracker.php [hunk 1]
```diff
@@ -30,7 +34,7 @@ class WC_Tracker {
 	/**
 	 * Hook into cron event.
 	 */
-	public static function init() {
+	public static function init() { // phpcs:ignore WooCommerce.Functions.InternalInjectionMethod.MissingFinal, WooCommerce.Functions.InternalInjectionMethod.MissingInternalTag -- Not an injection.
```
Comment (menakas) [inline plugins/woocommerce/includes/class-wc-tracker.php:37]: Thanks for adding these annotations. Useful addition. Much appreciate 👍

File: plugins/woocommerce/includes/class-wc-tracker.php [hunk 2]
```diff
@@ -45,10 +49,15 @@ public static function send_tracking_data( $override = false ) {
 			return;
 		}
 
+		/**
```
Comment (menakas) [inline plugins/woocommerce/includes/class-wc-tracker.php:52]: Thanks for the doc comments. Again, much useful and much appreciated.

File: plugins/woocommerce/includes/class-wc-tracker.php [hunk 3]
```diff
@@ -381,10 +410,9 @@ public static function get_product_counts() {
 	 * @return array
 	 */
 	private static function get_order_counts() {
-		$order_count      = array();
-		$order_count_data = wp_count_posts( 'shop_order' );
+		$order_count = array();
 		foreach ( wc_get_order_statuses() as $status_slug => $status_name ) {
-			$order_count[ $status_slug ] = $order_count_data->{ $status_slug };
+			$order_count[ $status_slug ] = wc_orders_count( $status_slug );
```
Comment (menakas) [inline plugins/woocommerce/includes/class-wc-tracker.php:415]: Would this count only those orders with the `shop_order` type for the status? What happens in the case of `shop_refund`? I am not sure if any refunded orders would show any difference in data.
These are my initial thoughts.
I shall check that soon.

File: plugins/woocommerce/includes/class-wc-tracker.php [hunk 4]
```diff
@@ -413,33 +441,59 @@ private static function get_orders() {
...
 
-		$gross_total = $wpdb->get_var(
+		$orders_table = OrdersTableDataStore::get_orders_table_name();
+
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+			$gross_total = $wpdb->get_var(
+				"
+				SELECT SUM(total_amount) AS 'gross_total'
+				FROM $orders_table
```
Comment (menakas) [inline plugins/woocommerce/includes/class-wc-tracker.php:451]: Glad to see those expensive joins gone 💥

File: plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php [hunk 5]
```diff
@@ -64,4 +67,55 @@ function( $pre, $args, $url ) use ( &$posted_data ) {
 		$this->assertArrayHasKey( 'wc_admin_disabled', $tracking_data );
 		$this->assertEquals( 'no', $tracking_data['wc_admin_disabled'] );
 	}
+
+	/**
+	 * @testDox Test orders tracking data.
+	 */
+	public function test_get_tracking_data_orders() {
```
Comment (menakas) [inline plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php:74]: Thanks for adding these tests. Long overdue item.

**Review by menakas — APPROVED** [review]
Thanks for the work to add HPOS compatibility for the tracker. 
I did some basic testing, and it appears fine. 

But I'd like to do some extended testing - especially on the `shop_refund` orders that I mentioned earlier. I shall do that and post if there are any problems.

For now, I'm glad to approve it.

File: plugins/woocommerce/includes/class-wc-tracker.php (hunk 3 above)
Comment (vedanshujain) [reply plugins/woocommerce/includes/class-wc-tracker.php:415]: Good catch, currently, it excludes shop order refunds currently, [as they are marked to not be counted.](https://github.com/woocommerce/woocommerce/blob/953a8f5c30c9041cf5d9a2eeac0ab7385a91b16d/plugins/woocommerce/includes/class-wc-post-types.php#L440) However, in the earlier code, we are not counting them either, is this expected?

Another notable difference between order counting is that, earlier we were only counting posts with `shop_order` type, but now we will also count order types that are registered with the count param set to true.

Comment (menakas) [reply plugins/woocommerce/includes/class-wc-tracker.php:415]: Thanks for the pointer. Yes, we're not counting in the earlier code. If they're not counted, it should be fine. I did one test with a refunded order and found that in both the cases, the count was the same.

**Review by jorgeatorres — COMMENTED** [review]
LGTM. Thank you @vedanshujain!

Note: not merging yet as this is still to be reviewed by @deltaWhiskey.

**Review by jorgeatorres — APPROVED** [review]

**Review by deltaWhiskey — APPROVED** [review]
Thanks for making tidying up the surrounding code while you were at it!

The new output is consistent with the old. Looks good.

The new HPOS code is definitely a performance win. Snapshot generation with the new logic takes about 70% the time that the old logic took, testing with 6,100 orders.

**Comment by vedanshujain** [comment]
Thanks for all the reviews folks, merging!

**Comment by lkraav** [comment]
These tracker queries absolutely murder our database on a 10M records posts table + 30M postmeta table. `SELECT gateway, currency, SUM(total) AS totals, COUNT(order_id) AS counts FROM ( SELECT orders.id AS order_id, MAX(CASE WHEN meta_key = '_payment_method' ...` query is taking 40+ min to complete.

How can we turn this tracker class off completely? Thanks!

**Comment by jorgeatorres** [comment]
@lkraav: You can turn off the tracker class completely by unchecking the "Enable tracking" setting in WC > Settings > Advanced > Woo.com. Alternatively, you could just set the `woocommerce_allow_tracking` option to `no` using WP-CLI.

**Pull request merged**